src/xma/src/xmaapi/xmaapi.cpp

```diff
@@ -267,6 +271,7 @@ void xma_thread2(uint32_t hw_dev_index) {
             xrt_device_obj.get_handle()->exec_wait(100);
         }
 
+        std::lock_guard<std::mutex> lock(g_xma_singleton->m_mutex);
```

Comment:
Would you want to protect the cpu_mode check above? It is not atomic. It may be worth just updating that variable to be atomic too.

Reply:
I'm not that familiar with the codebase and I'm not sure what the `cpu_mode` check really does. I guess it did not affect my issue.

Reply:
No worries, the reason I pointed it out is I see the mutex lock you added as a check on the data inside the singleton. Mainly you use it to defend the vector of XmaSessions which makes perfect sense.

However, in `struct XmaSingleton` (xmaapi.h:41) the cpu_mode is not defined as atomic.

```
typedef struct XmaSingleton
{
    XmaHwCfg          hwcfg;
    bool              xma_initialized;
    bool              kds_old;
    uint32_t          cpu_mode;
    std::mutex            m_mutex;
    std::atomic<uint32_t> num_decoders;
 ... More atomics ...
    std::atomic<uint32_t> num_of_sessions;
    std::vector<XmaSession> all_sessions_vec;// XMASessions
...
} XmaSingleton;
```

I believe we should use the mutex to defend access to all non-atomic variables of the struct.

Alternatively, you could make `cpu_mode` an atomic value which ensure its value in the check. And extending that change to all of the primitive types (except the `reserved` member variable) is a good idea since this is struct use between multiple threads.

Reply:
The variable is not atomic but it is not being changed in other threads, it is only set once in `xma_initialize` so the if statement here is secure, no potential race conditions.